pymc3/distributions/distribution.py

```diff
@@ -56,17 +56,32 @@ def __new__(cls, name, *args, **kwargs):
                             "a 'with model:' block, or use the '.dist' syntax "
                             "for a standalone distribution.")
 
-        if isinstance(name, string_types):
-            data = kwargs.pop('observed', None)
-            cls.data = data
-            if isinstance(data, ObservedRV) or isinstance(data, FreeRV):
-                raise TypeError("observed needs to be data but got: {}".format(type(data)))
-            total_size = kwargs.pop('total_size', None)
-            dist = cls.dist(*args, **kwargs)
-            return model.Var(name, dist, data, total_size)
-        else:
+        if not isinstance(name, string_types):
             raise TypeError("Name needs to be a string but got: {}".format(name))
 
+        data = kwargs.pop('observed', None)
+        cls.data = data
+        if isinstance(data, ObservedRV) or isinstance(data, FreeRV):
+            raise TypeError("observed needs to be data but got: {}".format(type(data)))
+        total_size = kwargs.pop('total_size', None)
+
+        dims = kwargs.pop('dims', None)
+        has_shape = 'shape' in kwargs
+        shape = kwargs.pop('shape', None)
+        if dims is not None:
+            if shape is not None:
+                raise ValueError("Specify only one of 'dims' and 'shape'")
+            if isinstance(dims, string_types):
+                dims = (dims,)
+            shape = model.shape_from_dims(dims)
+
+        # Some distribution do not accept shape=None
+        if has_shape or shape is not None:
```

Comment:
Why not just check on `shape is not None` and get rid of `has_shape`, which is used only here?

Reply:
I think it's to make sure that if `shape=None` is passed by the user (== present in kwargs) it is still passed to `cls.dist`

Reply:
Indeed, Adrian explained it to me below. Closing the comment.